content/en/blog/2023-02-07-building-a-raft-part-3.md

```diff
@@ -0,0 +1,115 @@
+    int32 lastLogTerm = 5;    // The term number of candidate’s last log entry
+}
+```
+
+In return, each voter will respond with a `VoteReply`, which contains their own unique identifier, their perspective of the term, and their vote.
+
+```proto
+message VoteReply {
+    string id = 1;            // ID of the Raft node sending the vote.
+    int32 term = 2;           // Current term number
+    bool voteGranted = 3;     // True if the voter has voted for the candidate
+}
+```
+
+#### 1. Vote "no" if the candidate is behind
+
+Just like in `AppendEntries`, we first need to check if the candidate has an up-to-date term. If the candidate requesting the vote has a term less than the recipient's term, that candidate is not up-to-date with the rest of the cluster, and should be denied the vote. On the other hand if the candidate has a term greater than ours, we should ensure that we are a follower and reset our election timeout since we know there is a node in the cluster more up to date. 
+
+#### 2. Vote "yes" if the candidate is ahead or even (but only vote once!)
+
+Each server in a Raft cluster keeps track of who it has most recently voted for (until the next successful heartbeat). If this `votedFor` field is null, then it hasn’t voted for anyone in the current term and is free to vote for the candidate that sent the request.
+
+Here's the full logic for the `RequestVote` function:
+
+```golang
+func (s *RaftServer) RequestVote(ctx context.Context, in *api.VoteRequest) (out *api.VoteReply, err error) {
+	s.Lock()
+	defer s.Unlock()
+	out = &api.VoteReply{Id: s.id, Term: s.currentTerm, VoteGranted: false}
+	lastLogIndex, lastLogTerm := s.lastLogIndexAndTerm()
+
+	// Check if the incoming request has a higher term than ours, if so then there is a node in the cluster 
+	// more up to date, so we ensure we are a follower and reset our election timeout
+	// Note: It's important that we return this node's current term so that the requester can check if it is 
+	// 		 out of date and if so, do the same (revert to follower and reset it's election timeout)
+	if in.Term > s.currentTerm {
+		fmt.Printf("RequestVote: in.Term > s.currentTerm, reverting to follower\n")
+		s.becomeFollower(in.Term)
+	}
+
+	// This (very complicated) check is to make sure the following things are true before granting the vote:
+	//	1. This nodes current term is the same as the requester's term
+	//  2. This node hasn't voted for another candidate
+	//  3. The lastLogTerm and LastLogIndex of the requestor is at least up to date with this node's
+	if s.currentTerm == in.Term && (s.votedFor == in.CandidateId || s.votedFor == "") &&
+		(int(in.LastLogTerm) > lastLogTerm || (in.LastLogTerm == int32(lastLogTerm) &&
+			in.LastLogIndex >= int64(lastLogIndex))) {
+		fmt.Printf("granting vote to %v\n", in.CandidateId)
+		out.VoteGranted = true
+		s.votedFor = in.CandidateId
+		s.lastHeartbeat = time.Now()
+	} 
+	return out, nil
+}
+```
```

Comment:
I think it would be nice to also have a sentence or two which summarizes the logic at a very high level to tie it all together. For example, some questions I might have as a reader:

1. Why is it important that we revert to a follower if we are out of date? What would happen if we remained a leader?
2. What is the purpose of checking the `votedFor` (e.g. prevent double votes) and the term/index of the requester?

Even if the answers seem obvious I think it's still beneficial to have them spelled out for the reader.

Reply:
That's a good idea. I added a summary paragraph after this code block answering these questions, let me know what you think.